### PR TITLE
Add Scheduler

### PR DESCRIPTION
**Pull request by Nacioszeczek**
Resolves PaperMC/docs#102

### CODE REVIEW

File: docs/paper/dev/api/scheduler.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+In Minecraft's case the amount of ticks per second is 20, meaning that the game loop is executed 20 times per second.
+With some math we can see that 1 tick is equal to 50 milliseconds. A tick taking more than 50ms to execute is the moment
+when your server starts to fall behind on its work and lag.
+
+### Converting between SI units and Minecraft ticks
+
+Every method of the scheduler that takes a delay or period uses ticks as a unit of time.
+
+:::warning
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:22]: I don’t think this warrants a warning. Maybe just info

File: docs/paper/dev/api/scheduler.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+:::
+
+Converting from human units to ticks and back is as simple as:  
+`ticks = seconds * 20`  
+`seconds = ticks / 20`
+
+:::note
+
+In Java and Kotlin to not lose precision you must use `ticks / 20.0`.
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:35]: I don’t think this is the place to explain basic java

**Review by olijeffers0n — CHANGES_REQUESTED** [review]
I’m assuming this is still a WIP due to it ending abruptly and it being a draft PR, but here are a couple comments.

File: docs/paper/dev/api/scheduler.md (hunk 1 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:22]: IMO time units for async methods would be somewhat expected to be milliseconds (when there's no `TimeUnit` param), however they are not therefore the big red warning

File: docs/paper/dev/api/scheduler.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+## What is a tick?
+
+Every game runs something called a game loop which essentially executes all the logic of the game over and over,
+a single execution of that loop is called a 'tick'.
+
+In Minecraft's case the amount of ticks per second is 20, meaning that the game loop is executed 20 times per second.
+With some math we can see that 1 tick is equal to 50 milliseconds. A tick taking more than 50ms to execute is the moment
+when your server starts to fall behind on its work and lag.
+
+### Converting between SI units and Minecraft ticks
```
[the remark this answers is not in the document]
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:18]: Missed when changing to human units as below
>Converting from human units to ticks and back is as simple as:  
agreed & changed

File: docs/paper/dev/api/scheduler.md (hunk 2 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:35]: considering the IDE does display a warning when the variable's type is `float`/`double` and above is pseudo-code, agreed

File: docs/paper/dev/api/scheduler.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+:::warning
+
+The above also applies to the asynchronous methods. If you need a scheduler with millisecond precision, create your own
+[`ScheduledExecutorService`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html).
+
+:::
+
+Converting from human units to ticks and back is as simple as:  
+`ticks = seconds * 20`
```
[the remark this answers is not in the document]
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:30]: I don't see how secondsPassed is better, imo readable as is

File: docs/paper/dev/api/scheduler.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+`TimeUnit.SECONDS.toMinutes(ticks / 20)`
+
+## Obtaining the scheduler
+
+To obtain a scheduler you can use the instance method on the `Server` class, e.g. in your `onEnable` method:
+
+```java
+@Override
+public void onEnable() {
+	BukkitScheduler scheduler = this.getServer().getScheduler();
```
[the remark this answers is not in the document]
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:52]: convention about `this` was already broken with my previous PR of Custom InventoryHolders, will leave for staff to decide

File: docs/paper/dev/api/scheduler.md [hunk 6]
```diff
@@ -51,4 +51,89 @@ To obtain a scheduler you can use the instance method on the `Server` class, e.g
...
+
+## Scheduling tasks
+
+Scheduling a task requires you to pass:
+- your plugin's instance,
+- the code to run, either with a `Runnable` or `Consumer<BukkitTask>` (the differences and usage are explained below),
+- the delay in ticks before the task should run,
+- the period in ticks between each execution of the task if you're scheduling a repeating task.
+
+### Using `Runnable`
```
Comment (Nacioszeczek) [inline docs/paper/dev/api/scheduler.md:58]: not so glad of the structure / naming of this section

**Comment by Machine-Maker** [comment]
A large part of the scheduler is being able to do things off the main thread. But this is barely explained in the PR. Is that intentional? Cause I feel like talking about "async" here makes a lot of sense.

**Comment by Machine-Maker** [comment]
Yeah, I think the concept of "async" isn't very well understood, along with its limitations when interacting with "world state". But sounds like you have plans, I just think in addition to example sections, there should be a blurb on what "async" means in this context.

**Comment by olijeffers0n** [comment]
Yep, as MM says explaining the concept behind the event loop and the pros/cons async is vital.

I think the page name could be better in the sidebar. Even if it is just “The Scheduler” it would be better as just “Scheduler” looks weird IMO.

**Comment by Nacioszeczek** [comment]
@Machine-Maker I assume the BukkitScheduler executes the tasks at the start/end of a tick therefore is still affected if mspt exceeds 50ms, and your own `ScheduledExService` wouldn't care at all?
I just think `ScheduledExecutorService` is worth mentioning *somewhere* in this doc as an alternative

**Comment by Nacioszeczek** [comment]
@olijeffers0n wanna elaborate? Not sure I understood what you mean
> the concept behind the event loop

File: docs/paper/dev/api/scheduler.md [hunk 7]
```diff
@@ -184,20 +184,34 @@ public class MyDynamicTask implements Runnable {
...
+we use the `runTaskLater` method too and not `runTaskTimer`.
+
+:::
+
+:::info
+
+We use a try-finally block to ensure the task is scheduled again. Otherwise if the methods before scheduling
+the tasks would throw an exception the task would never be scheduled again.
+
+:::
```
Comment (Nacioszeczek) [inline docs/paper/dev/api/scheduler.md:217]: not sure which warrants a note and which an info? opinions very much welcome

Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:217]: guess note reiterates sth that can be missed and info explains stuff 🤷🏻

**Comment by Nacioszeczek** [comment]
Waiting for any suggestions, some criticism, etc. since I'm somewhat out of ideas for this;

also making ready to review since IMHO asynchronous tasks examples are not really needed since that would be just repeating the synchronous examples with different method names

**Comment by olijeffers0n** [comment]
Looks good, is it worth mentioning bukkit runnables?

**Comment by Nacioszeczek** [comment]
All scheduler methods that take `BukkitRunnable` are deprecated and the methods on `BukkitRunnable` itself I'm not a fan of because calling static stuff

File: docs/paper/dev/api/scheduler.md [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+The `BukkitScheduler` can be used to schedule your code to be run later or run it repeatedly.
+
+## What is a tick?
+
+Every game runs something called a game loop which essentially executes all the logic of the game over and over,
+a single execution of that loop is called a 'tick'.
+
+In Minecraft's case the amount of ticks per second is 20, meaning that the game loop is executed 20 times per second.
+With some math we can see that 1 tick is equal to 50 milliseconds. A tick taking more than 50ms to execute is the moment
```
Comment (kennytv) [inline docs/paper/dev/api/scheduler.md:15]:
```suggestion
In Minecraft's case the amount of ticks per second is 20, or one tick every 50 milliseconds,
meaning that the game loop is executed 20 times per second. A tick taking more than 50ms to execute is the moment
```

**Review by kennytv — CHANGES_REQUESTED** [review]

File: docs/paper/dev/api/scheduler.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Machine-Maker) [reply docs/paper/dev/api/scheduler.md:23]: I think its fine as is.

File: docs/paper/dev/api/scheduler.md [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+```
+
+Or use a lambda expression which is great for simple and short tasks:
+
+```java
+scheduler.runTaskLater(
+	plugin, 
+	// Lambda:
+	() -> {
+		this.plugin.getServer().broadcast(Component.text("Hello, World!")
```
Comment (Machine-Maker) [inline docs/paper/dev/api/scheduler.md:120]: Missing a closing `)`

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by Nacioszeczek** [comment]
Gonna get to it soon

File: docs/paper/dev/api/scheduler.md [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+This task will run repeatedly, first time after 10 seconds, and then every 5 seconds. After 10 minutes
+the task will be canceled entirely.
+
+```java
+BukkitTask task = scheduler.runTaskTimer(plugin, () -> {
+	server.broadcast(Component.text("Hello, World!"));
+}, 10 * 20, 5 * 20);
+
+scheduler.runTaskLater(plugin, () -> task.cancel(), TimeUnit.MINUTES.toSeconds(10) * 20);
```
[the remark this answers is not in the document]
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:197]: Not sure that's what you meant but changed

File: docs/paper/dev/api/scheduler.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
+﻿---
+slug: /dev/scheduler
+---
+
+# Task Scheduler
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:5]: Why is the page called task scheduler if you immediately refer to it with the `BukkitScheduler` name. Maybe just name the page "The Bukkit Scheduler" or something along those lines?

File: docs/paper/dev/api/scheduler.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+The `BukkitScheduler` can be used to schedule your code to be run later or run it repeatedly.
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:7]:
```suggestion
The `BukkitScheduler` can be used to schedule your code to be run later or repeatedly.
```

File: docs/paper/dev/api/scheduler.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+}
+```
+
+## Scheduling tasks
+
+Scheduling a task requires you to pass:
+- your plugin's instance,
+- the code to run, either with a `Runnable` or `Consumer<BukkitTask>`,
+- the delay in ticks before the task should run,
+- if you're scheduling a repeating task - the period in ticks between each execution of the task.
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:49]:
```suggestion
- Your plugin's instance,
- The code to run, either with a `Runnable` or `Consumer<BukkitTask>`,
- The delay in ticks before the task should run for the first time,
- If you're scheduling a repeating task - the period in ticks between each execution of the task.
```

File: docs/paper/dev/api/scheduler.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+```java
+scheduler.runTaskLater(
+	plugin, 
+	// Lambda:
+	() -> {
+		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
+	},
+	// End of the lambda
+	20);
```
Comment (olijeffers0n) [inline docs/paper/dev/api/scheduler.md:123]:
```suggestion
	plugin, /* Lambda: */ () -> {
		this.plugin.getServer().broadcast(Component.text("Hello, World!"));
	}, /* End of the lambda */ 20);
```

Keeps it styled the same as the next example and looks cleaner

**Review by olijeffers0n — CHANGES_REQUESTED** [review]
Looks like its pretty much there, just a few suggestions really.

File: docs/paper/dev/api/scheduler.md (hunk 11 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:5]: I'd prefer something like `Scheduling tasks` over including Bukkit in the name

File: docs/paper/dev/api/scheduler.md (hunk 13 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:49]: Think of it as a long sentence, hence the commas and the period at the end,
will add `for the first time`

File: docs/paper/dev/api/scheduler.md (hunk 12 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:7]: Will change

File: docs/paper/dev/api/scheduler.md (hunk 14 above)
Comment (Nacioszeczek) [reply docs/paper/dev/api/scheduler.md:123]: Will change

File: docs/paper/dev/api/scheduler.md [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+```
+```java
+scheduler.runTaskTimer(plugin, new MyConsumerTask(someEntity), 0, 20);
+```
+
+Or use a lambda expression which again is much cleaner for short and simple tasks:
+
+```java
+scheduler.runTaskTimer(plugin, /* Lambda: */ task -> {
+	if(this.entity.isDead()) {
```
Comment (Lulu13022002) [inline docs/paper/dev/api/scheduler.md:162]: Missing a space between `if` and `(`

**Comment by Nacioszeczek** [comment]
succeeded by #204

**Pull request closed without merging**